lib/backend/k8s/resources/watcher.go

```diff
@@ -155,56 +155,67 @@ func (crw *k8sWatcherConverter) processK8sEvents() {
 func (crw *k8sWatcherConverter) convertEvent(kevent kwatch.Event) []*api.WatchEvent {
 	var kvps []*model.KVPair
 	var err error
-	if kevent.Type != kwatch.Error {
-		k8sRes := kevent.Object.(Resource)
-		kvps, err = crw.converter(k8sRes)
-		if err != nil {
-			crw.logCxt.WithError(err).Warning("Error converting Kubernetes resource to Calico resource")
-			return []*api.WatchEvent{{
-				Type:  api.WatchError,
-				Error: err,
-			}}
-		}
 
-		if len(kvps) == 0 {
-			return nil
+	switch kevent.Type {
+	case kwatch.Error:
+		// An error directly from the k8s watcher is a terminating event.
+		return []*api.WatchEvent{{
+			Type: api.WatchError,
+			Error: cerrors.ErrorWatchTerminated{
+				Err: fmt.Errorf("terminating error event from Kubernetes watcher: %v", kevent.Object),
+			},
+		}}
+	case kwatch.Deleted:
+	case kwatch.Added:
+	case kwatch.Modified:
+	default:
+		return []*api.WatchEvent{{
+			Type:  api.WatchError,
+			Error: fmt.Errorf("unhandled Kubernetes watcher event type: %v", kevent.Type),
+		}}
+	}
+
+	k8sRes := kevent.Object.(Resource)
+	kvps, err = crw.converter(k8sRes)
+	if err != nil {
+		crw.logCxt.WithError(err).Warning("Error converting Kubernetes resource to Calico resource")
+		return []*api.WatchEvent{{
+			Type:  api.WatchError,
+			Error: err,
+		}}
+	}
+
+	if len(kvps) == 0 {
+		return nil
+	}
+
+	return crw.buildEventsFromKVPs(kvps, kevent.Type)
+}
+
+func (crw *k8sWatcherConverter) buildEventsFromKVPs(kvps []*model.KVPair, t kwatch.EventType) []*api.WatchEvent {
+	var getEvent func(*model.KVPair) *api.WatchEvent
+	switch t {
```

Comment:
This switch seems a little redundant, you could set the get even in the above switch (since you already have the same cases). With that in mind, you could just remove this function loop through the kvps at the end of the previous function.

Reply:
Yeah, I originally suggested a split like this in my slack chat w/ Erik.

I was imagining it would be called inline the above switch statement, but see now that it's sorta tricky since we don't build the events the same for each type. I'm also OK with re-combining these two functions and just looping over the events like we used to.

Reply:
The intention for not doing it in the above switch is to de-couple handling of special events - e.g., Errors and Unexpected types.

That was sorta the bug in the first place - we were handling errors in the same loop as we were handling other update types, and didn't spot the pre-reqs to reach that loop in the first place.

Reply:
Makes sense